service/share/share.go

```diff
@@ -0,0 +1,124 @@
+// Share is
+type Share = namespace.PrefixedData8
+
+// Service provides as simple interface to access any DataSquare/Block Share on the network.
+//
+// All Get methods follow the following flow:
+// 	* Check local storage for the requested Share.
+// 		* If exist
+// 			* Load from disk
+//			* Return
+//  	* If not
+//  		* Find provider on the network
+//      	* Fetch the Share from the provider
+//			* Store the Share
+//			* Return
+type Service interface {
+	// GetShare loads a Share committed to the given DataAvailabilityHeader by its Row anc Col coordinates in the
+	// erasures data square or block.
+	GetShare(ctx context.Context, dah header.DataAvailabilityHeader, row, col int) (Share, error)
+
+	// GetShares loads all the Shares committed to the given DataAvailabilityHeader as a 2D array/slice.
+	// It also optimistically executes erasure coding recovery.
+	GetShares(context.Context, header.DataAvailabilityHeader) ([][]Share, error)
+
+	// GetSharesByNamespace loads all the Shares committed to the given DataAvailabilityHeader as a 1D array/slice.
+	GetSharesByNamespace(context.Context, header.DataAvailabilityHeader, namespace.ID) ([]Share, error)
+
+	// Starts the Service.
+	Start(context.Context) error
+
+	// Stops the Service.
+	Stop(context.Context) error
+}
+
+// NewService creates new basic share.Service.
+func NewService(dag format.DAGService) Service {
+	return &service{
+		dag: dag,
+	}
+}
+
+type service struct {
+	dag format.DAGService
+	// session is dag sub-session that applies optimization for fetching/loading related nodes, like shares
+	// prefer session over dag for fetching nodes.
+	session format.NodeGetter
+	// cancel controls lifecycle of the session
+	cancel context.CancelFunc
+}
+
+func (s *service) Start(context.Context) error {
+	if s.session != nil || s.cancel != nil {
+		return nil
+	}
+
+	// NOTE: The ctx given as param is used to control Start flow and only needed when Start is blocking,
+	// but this one is not.
+	//
+	// The newer context here is created to control lifecycle of the session.
+	ctx, cancel := context.WithCancel(context.Background())
+	s.cancel = cancel
+	s.session = merkledag.NewSession(ctx, s.dag)
+	return nil
+}
+
+func (s *service) Stop(context.Context) error {
+	if s.session == nil || s.cancel == nil {
+		return nil
+	}
+
+	s.cancel()
+	s.cancel = nil
+	s.session = nil
+	return nil
+}
+
+func (s *service) GetShare(ctx context.Context, dah header.DataAvailabilityHeader, row, col int) (Share, error) {
+	rootCid, err := plugin.CidFromNamespacedSha256(dah.RowsRoots[row])
+	if err != nil {
+		return nil, err
+	}
+
+	nd, err := ipld.GetLeaf(ctx, s.dag, rootCid, col, len(dah.ColumnRoots))
+	if err != nil {
+		return nil, err
+	}
+
+	// we exclude one byte, as it is not part of the share, but encoding of IPLD NMT Node type.
+	// TODO(@Wondertan): There is way to understand node type indirectly,
+	//  without overhead(storing/fetching) of appending it to share.
+	return nd.RawData()[1:], nil
```

Comment:
nit and not necessary specifically to this PR, but it will be cleaner for my impl of GetSharesByNamespace too is if we create a method to get the raw data without the prepended byte bc it will just look cleaner and easier to understand.

maybe i will add it in my PR for get shares by namespace

Reply:
I agree that cleaner would be without seeing this prepended byte. The reason I kept it here is visibility that this should be fixed and not just by hiding this byte prefix, but to encourage everyone to remove it completely from our data structure as it is redundant. Thus, I would like to keep this ugliness to remind us to fix it completely.

Reply:
Yeah would be more preferable to remove the prepended byte entirely.